Add/35126 ces exit prompt orders

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This adds order exit tracking, to show a notice when user exits the order page when having made changes without updating them.
Part of #35126  .

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Make sure you have tracks enabled and also outputted in the console: localStorage.setItem( 'debug', 'wc-admin:*' );
3. Go to **WooCommerce > Orders** and select an existing order or create a new one, and make some changes
4. Now go to another page like **WooCommerce > Home** and select **Leave** if it warns you about unsaved changes (it may not).
5. A notice should show on the new page that allows you to share feedback, something to the affect of `How easy or difficult was it to update this order?`
6. Click the share feedback and submit feedback, notice how the `wcadmin_ces_snackbar_view`, `wcadmin_ces_view`, and `wcadmin_ces_feedback` include the `ces_location` prop with the  `outside` value.
7. For the CES modal check if the copy matches that listed in this document for the Edit order usage action: [copy](https://docs.google.com/spreadsheets/d/1t0JplRZL3cC3vNCATmMy95surJzYH_uhsitznelYcRI/edit?usp=sharing).
You may also want to try dismissing the notice and see if the prop has been added to `wcadmin_ces_snackbar_dismiss`
8. Now delete the `woocommerce_ces_shown_for_actions` option and disable tracking under **WooCommerce > Settings > Advanced > Woocommerce.com**
9. Make sure the notice doesn't show up for the above anymore and that `customer-effort-score-exit-page` in local storage does not get updated, but stays empty.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add exit tracking for orders

louwie17 committed:
Update exit page CES action

louwie17 committed:
Fix order hook name

octaedro commented:
Hey @louwie17, 
Since this PR is still a `Draft`, I changed its status from `In review` to `In progress`.

louwie17 committed:
Add changelog

louwie17 commented:
> Hey @louwie17, Since this PR is still a `Draft`, I changed its status from `In review` to `In progress`.

Thanks @octaedro, I did just move it back to `in review` again, as I just rebased it and it should be ready for review.

joelclimbsthings commented on plugins/woocommerce-admin/client/wp-admin-scripts/order-tracking/exit-order-page.ts:
```diff
@@ -0,0 +1,51 @@
+	post?: HTMLFormElement;
+} & HTMLCollectionOf< HTMLFormElement >;
+const forms: FormElements = document.forms;
+if ( forms && forms.post ) {
```
Is there a reason not to do `if( forms?.post ) {` here?

louwie17 replied:
No, that seems like a better approach, thanks!

joelclimbsthings commented on plugins/woocommerce-admin/client/wp-admin-scripts/order-tracking/exit-order-page.ts:
```diff
@@ -0,0 +1,51 @@
+					? JSON.stringify( newFormData[ key ] )
+					: newFormData[ key ];
+			if ( value !== newValue ) {
+				isDirty = true;
```
It might be a little cleaner to get rid of the `isDirty` variable, and just `return true` here, and `false` below. You also won't need the `break;`

joelclimbsthings reviewed (dismissed):
Great work @louwie17 , working well! Just had a couple very minor comments and I'll pre-approve.

louwie17 committed:
Address PR feedback

louwie17 commented:
> Great work @louwie17 , working well! Just had a couple very minor comments and I'll pre-approve.

Thanks @joelclimbsthings, I updated this PR with your suggested changes, if you have time for a re-approval.

joelclimbsthings reviewed (approved):
Looks great! Thanks @louwie17 :shipit:
